### PR TITLE
[FIXED] Interest policy and staggered filtered consumers could fail to remove messages.

### LINKED ISSUE #2873: Interest policy: nats server won't delete acked messages if consumers are not keeping up
## Defect
when using Interest policy, nats server won't delete acked messages if consumers are not keeping up

Make sure that these boxes are checked before submitting your issue -- thank you!

 - [x] Included `nats-server -DV` output - unfortunately I have no access atm, please let me know if you need it
 - [x] Included a [Minimal, Complete, and Verifiable example] (https://stackoverflow.com/help/mcve)

#### Versions of `nats-server` and affected client libraries used:
nats-server 2.7.2
python nats client 2.0.0
#### OS/Container environment:
docker container nats 2.7.2 alpine

#### Steps or code to reproduce the issue:
1) create jetstream with following config
```
Configuration:

             Subjects: messages.*
     Acknowledgements: true
            Retention: File - Interest
             Replicas: 1
       Discard Policy: New
     Duplicate Window: 2m0s
    Allows Msg Delete: true
         Allows Purge: true
       Allows Rollups: false
     Maximum Messages: unlimited
        Maximum Bytes: unlimited
          Maximum Age: unlimited
 Maximum Message Size: unlimited
    Maximum Consumers: unlimited
```
2. create two polling consumers, one for messages.msg1 and another for messages.msg2

```
Configuration:

        Durable Name: MSG_XXX_CONSUMER
           Pull Mode: true
      Filter Subject: messages.message1
      Deliver Policy: All
          Ack Policy: Explicit
            Ack Wait: 30s
       Replay Policy: Instant
     Max Ack Pending: 20,000
   Max Waiting Pulls: 512
```

3. consumers are slower than inbound producer but ack messages normally within configured timeframe
4. messages get stacked up inside nats server filesystem, waiting for consumer to pick them up
5. stop producer

#### Expected result:
messages should be eventually all consumed, acked and removed from nats file system

#### Actual result:
messages are all consumed, acked, but stay on nats file system.
when we delete durable consumers from server and reregister consumers without producing any messages on stream, messages eventually disappear from nats server

Screenshot 2022-02-16 at 13 21 48

### PR DESCRIPTION
**Pull request by derekcollison**
On filtered subjects when consumers were staggered we need to disqualify a filtered consumer as not applicable before doing any sequence and ack floor determinations are made to see if we can delete a message from a stream.

Signed-off-by: Derek Collison <derek@nats.io>

Resolves #2873

/cc @nats-io/core

### CODE REVIEW

File: server/consumer.go [hunk 1]
```diff
@@ -2085,6 +2085,16 @@ func (o *consumer) needAck(sseq uint64) bool {
 	var asflr, osseq uint64
 	var pending map[uint64]*Pending
 	o.mu.RLock()
+
+	// Check first if we are filtered, and if so check if this is even applicable to us.
+	if o.isFiltered() && o.mset != nil {
+		subj, _, _, _, err := o.mset.store.LoadMsg(sseq)
+		if err != nil || !o.isFilteredMatch(subj) {
+			o.mu.RUnlock()
```
Comment (kozlovic) [inline server/consumer.go:2093]: Oops, I see that you already have unlock the consumer line 2087. So move o.mu.RLock() from line 2087 down after the outer if, assuming that o.isFiltered() or o.isFilteredMatch() are not grabbing the lock.

**Review by kozlovic — CHANGES_REQUESTED** [review]

File: server/consumer.go (hunk 1 above)
Comment (derekcollison) [reply server/consumer.go:2093]: I think its ok.

Comment (kozlovic) [reply server/consumer.go:2093]: Sorry, I read as an RUnlock() above...

**Review by kozlovic — APPROVED** [review]
LGTM

**Pull request merged**